torch/distributed/hooks.py

```diff
@@ -0,0 +1,138 @@
+import logging
+import os
+import threading
+import time
+from dataclasses import dataclass
+from typing import Callable, Dict, List, Optional
+
+import torch.distributed as dist
+import torch.distributed.distributed_c10d as c10d
+
+from torch._C._distributed_c10d import _dequeue_c10d_event, _enable_event_collection
+
+__all__ = [
+    "CollectiveStatus",
+    "COLLECTIVE_HOOK_TYPE",
+    "PG_HOOK_TYPE",
+    "register_collective_start_hook",
+    "register_collective_end_hook",
+    "register_process_group_hook",
+]
+
+
+@dataclass
+class CollectiveStatus:
+    pg_name: str = "unknown"  # This name matches the one informed in the pgreg cb
+    backend: str = "unknown"  # Name of the backend used
+    sequence_number: int = -1  # This name matches the one informed in the pgreg cb
+    operation: str = "unknown"  # collective name
+    timestamp: int = 0  # timestamp to the earliest time we noticed this event
+    duration: Optional[float] = None  # value in milliseconds it took executing
+
+
+COLLECTIVE_HOOK_TYPE = Callable[[CollectiveStatus], None]
+PG_HOOK_TYPE = Callable[[dist.ProcessGroup, str], None]
+
+"""
+TODO:
+    hook timing
+
+"""
+logger = logging.getLogger(__name__)
+_cb_thread: Optional[threading.Thread] = None
+_start_callbacks: List[COLLECTIVE_HOOK_TYPE] = []
+_end_callbacks: List[COLLECTIVE_HOOK_TYPE] = []
+_pp_r = -1
+_pp_w = -1
+
+
+def _c10d_pg_hooks_loops():
+    while True:
+        # we don't care about the result, this is how we implement notification
+        _ = os.read(_pp_r, 1)
```

Comment:
Do we really need this mechanism? You could write a pure C++ wrapper around `std::deque` which could provide methods like `queue.poll(timeout)` and `queue.read()`

Reply:
The original design was a wrapper around std::deque but using a blocking wait. 
This was problematic as it blocked python from shutting down - regardless if the thread was a daemon thread or not.


I tried using atexit callbacks to unstuck the background thread but that didn't work due to python shutdown behavior.

We could switch to using that design with a timed wait which hopefully would not happen and increase the number of context switches.

Reply:
It might make sense to switch to std::deque + timed wait once we have a c++ alternative as the performance implication of the python API would not matter as much.